### ISSUE TITLE
[question] Fish completions to maintain here or at fish-shell?

### ISSUE THREAD

**Issue by SanskritFritz** [borgbackup/borg#3086]
I wrote tab completion definitions for the fish-shell.
https://github.com/SanskritFritz/fish-shell/blob/patch-1/share/completions/borg.fish
My question is, would you like to include it into the borg distribution or should I merge it to fish? Fish does have a history of including completions for foreign packages, so that would not be a problem.
However, if borg includes the completions, changes in borg would be reflected sooner in the completion definitons. I'm willing to maintain it, even over at fish-shell.

**Comment by ThomasWaldmann** [comment]
See there:
https://github.com/borgbackup/borg/issues/2862

**Comment by SanskritFritz** [comment]
I'm closing this, since the decision has already been made.

**Comment by ThomasWaldmann** [comment]
@SanskritFritz See #9990 - can you test it please?

**Comment by SanskritFritz** [comment]
I have to pull up a container for testing. Can you please point me to an example?

**Comment by ThomasWaldmann** [comment]
Example for what precisely?

**Comment by SanskritFritz** [comment]
I want to test borg 2 installed alongside my prod borg. If I use borg2 the completions would still taken from the old files. How can I virtualize the environment for borg 2?

**Comment by ThomasWaldmann** [comment]
Maybe you could also use a different user account?

Then you could load the borg2 completions in your default shell rc files.

The borg2 installation can be in a virtual env, so it does not interfere with your system borg1.

The completions can be automatically created by `borg completion fish`:
- either once and write the output to some file you load
- every time the shell starts (bad: takes a short moment, good: updates completions automatically for an updated borg)

**Comment by SanskritFritz** [comment]
oh boy the simplest solution didn't occur to me...thanks, I'll test it soon

**Comment by SanskritFritz** [comment]
First test results. More testing will be done, but first a question: should I list the tests here of open a dedicated issue on github?

---
`borg2 --repo <tab>` or `borg2 --repo=<tab>` or `borg2 -r <tab>`

No path completions offered. IMO this is essential, especially if I don't remember the path exactly.

---
```
~/temp> borg2 --repo borg2testrepo repo-create
error: the following arguments are required: !!repo-create.encryption!!

~/temp> borg2 --repo borg2testrepo repo-create !!repo-create.encryption!!=repokey-aes-ocb
error: unrecognized arguments: !!repo-create.encryption!!=repokey-aes-ocb
```
Misleading error message for option (!! just for emphasis).

Also the example in [quickstart.html](https://borgbackup.readthedocs.io/en/2.0.0b20/quickstart.html) provides the key `repokey-aes-ocb` which doesn't exist in borg2 anymore.

---
```
~/temp> borg2 --repo borg2testrepo repo-create --encryption=<tab>
…cryption=none          (select cipher / AE algorithm: 'none', 'authenticated', 'aes256-ocb' or 'chacha20-poly1305' **(required)**)
…cryption=authenticated  (select cipher / AE algorithm: 'none', 'authenticated', 'aes256-ocb' or 'chacha20-poly1305' **(required)…)
…cryption=aes256-ocb    (select cipher / AE algorithm: 'none', 'authenticated', 'aes256-ocb' or 'chacha20-poly1305' **(required)**)
…cryption=chacha20-poly1305  (select cipher / AE algorithm: 'none', 'authenticated', 'aes256-ocb' or 'chacha20-poly1305' **(requi…)
```
The provided explanations for the encryptions are strange for me, the choices are already listed, so why list them again in the explanations? I'd rather expect a short and sweet explanation for the encryptions.

---
```
borg2 list --repo=borg2testrepo <tab>
```
I'd expect a list of names (the second positional argument) offered from that repo.

---

**Comment by ThomasWaldmann** [comment]
best would be to cross-check with https://github.com/tqdm/shtab - check if an issue or PR is already open there, if not, file one.

**Comment by SanskritFritz** [comment]
So, it seems I should first understand how the completions are generated via shtab, because I have to understand how to make shtab lists borg archives or even paths for example.

**Comment by SanskritFritz** [comment]
At this time (new job, really stresses me out) I don't have the energy to learn all that stuff. I would gladly test the completions, but I can't dig into the design details, sorry. So, it it's ok with you, I'll do further tests and report them.

**Comment by ThomasWaldmann** [comment]
Sure, thanks for helping!

**Comment by SanskritFritz** [comment]
Hello, I tested again with version 2.0.0b23. All of the above issues still persist.

**Comment by ThomasWaldmann** [comment]
@SanskritFritz should be fixed in #10188.

If you find more issues, please file new tickets.

**Comment by SanskritFritz** [comment]
OK, I'll test further.
